internal/virtualization/cmd/powerstate/vmop/vmop.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vmop
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/deckhouse/virtualization/api/client/kubeclient"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type VirtualMachineOperation struct {
+	client kubeclient.Client
+}
+
+func (v VirtualMachineOperation) Stop(ctx context.Context, vmName, vmNamespace string, wait, force bool) (msg string, err error) {
+	vmop := v.newVMOP(vmName, vmNamespace, v1alpha2.VMOPOperationTypeStop, force)
+	return v.do(ctx, vmop, wait)
+}
+
+func (v VirtualMachineOperation) Start(ctx context.Context, vmName, vmNamespace string, wait bool) (msg string, err error) {
+	vmop := v.newVMOP(vmName, vmNamespace, v1alpha2.VMOPOperationTypeStart, false)
+	return v.do(ctx, vmop, wait)
+}
+
+func (v VirtualMachineOperation) Restart(ctx context.Context, vmName, vmNamespace string, wait, force bool) (msg string, err error) {
+	vmop := v.newVMOP(vmName, vmNamespace, v1alpha2.VMOPOperationTypeRestart, force)
+	return v.do(ctx, vmop, wait)
+}
+
+func (v VirtualMachineOperation) do(ctx context.Context, vmop *v1alpha2.VirtualMachineOperation, wait bool) (msg string, err error) {
+	if wait {
+		vmop, err = v.createAndWait(ctx, vmop)
+	} else {
+		vmop, err = v.create(ctx, vmop)
+	}
+	msg = v.generateMsg(vmop)
+	return msg, err
+}
+
+func (v VirtualMachineOperation) generateMsg(vmop *v1alpha2.VirtualMachineOperation) string {
+	if vmop == nil {
+		return ""
+	}
+	key := types.NamespacedName{Namespace: vmop.GetNamespace(), Name: vmop.GetName()}
+	vmKey := types.NamespacedName{Namespace: vmop.GetNamespace(), Name: vmop.Spec.VirtualMachineName}
+	phase := vmop.Status.Phase
+
+	sb := strings.Builder{}
+	sb.WriteString(fmt.Sprintf("VirtualMachine %q ", vmKey.String()))
+
+	if v.isFinished(vmop) {
+		if !v.isCompleted(vmop) {
+			sb.WriteString("was not ")
+		}
+		switch vmop.Spec.Type {
+		case v1alpha2.VMOPOperationTypeStart:
+			sb.WriteString("started. ")
+		case v1alpha2.VMOPOperationTypeStop:
+			sb.WriteString("stopped. ")
+		case v1alpha2.VMOPOperationTypeRestart:
+			sb.WriteString("restarted. ")
+		}
+	} else {
+		sb.WriteString("in progress. ")
```

Comment:
It is much clearer now! But "in progress" message is noisy:

> VirtualMachine "my-vm" in progress. VirtualMachineOperation "my-vm-qwd23" was not finished. phase=InProgress.

I propose to add operation type and add descriptions for Pending and InProgress phases (see next comment):

> VirtualMachine "my-vm" stopping. VirtualMachineOperation "my-vm-qwd23" is in progress.

> VirtualMachine "my-vm" starting. VirtualMachineOperation "my-vm-qwd23" pending.


```suggestion
		switch vmop.Spec.Type {
		case v1alpha2.VMOPOperationTypeStart:
			sb.WriteString("starting. ")
		case v1alpha2.VMOPOperationTypeStop:
			sb.WriteString("stopping. ")
		case v1alpha2.VMOPOperationTypeRestart:
			sb.WriteString("restarting. ")
		}
```